[Step1] KangKyung

KangKyung opened the pull request:
@delmaSong 
delma! 안녕하세요!! 드디어 보내는 step1 PR입니다!!
시간이 제법 걸렸음에도, 아직 해결하지 못한 부분이 있어요😢
혼자 끙끙대는거 보다, 여쭤보면서 진행하는 것이 나을거 같아
도중에 PR을 보내게 되었습니다..!

## 구현 포인트
- 전체 UI를 **code base**로 구현하였습니다.
- **MVVM패턴**을 적용하였습니다.

## 작동 화면

![iPhone](https://user-images.githubusercontent.com/38858863/120751326-ce9f6c80-c542-11eb-9c34-914c05a94855.gif)
![iPad](https://user-images.githubusercontent.com/38858863/120751335-d19a5d00-c542-11eb-8da4-361c61322ea0.gif)

## 해결하지 못한 점

음... 제가 `Split View`를 제대로 활용하지 못하고 있는거 같아요😭
MVVM 패턴을 생각하면서, 넷상에 있는 Split View 만들기 코드들의 형태에 곧이곧대로 따르지않았는데
그로인해 문제들이 나오는거 같아요...

#### 기기가 아이폰인 경우, 화면전환이 안되는 문제

테이블 뷰에서, 셀을 클릭하면 해당 셀에 해당하는 뷰가 실행되지 않는 문제가 있습니다..
현재는

```swift
// FIXME: - 기기가 아이폰일 경우, 화면전환이 안되는 문제
if UIDevice.current.userInterfaceIdiom == .phone {
  listVC.navigationController?.pushViewController(memoVC, animated: true)
}
```

이런 식으로 하드코딩을 해뒀는데..
아무리 봐도 올바른 방법이 아닌거 같아, 고민 방향을 잡아주셨으면 좋겠어요😭

#### 기기가 아이폰인 경우, 가로화면 때 split 형태가 되지않는 문제

<img width="498" alt="Issue_iPhone" src="https://user-images.githubusercontent.com/38858863/120751372-da8b2e80-c542-11eb-9bf0-a518471bf8b0.png">

문제의 원인조차 찾지 못하고 있습니다..
아무래도, 앞서 말씀드린거 처럼 Split View를 잘 못쓰고있는게 아닐까 싶은데..
고민할 포인트를 잡아주시면 좀 더 열심히 고민해 보겠습니다..!

#### 기기아 아이패드인 경우, 세로화면에서 뒤로가기 버튼이 생성되지 않는 문제

<img width="436" alt="Issue_iPad" src="https://user-images.githubusercontent.com/38858863/120751394-e1b23c80-c542-11eb-82a3-42e9de63cf36.png">

이 역시 Split View를 잘 못쓰는 부분에서 파생된 문제같습니다😭😭😭

KangKyung committed:
Feat : table view code base 구현

- dummy data 출력

KangKyung committed:
Feat : memo view code base 구현

- table view cell click event 구현
- memo view를 navigation으로 연결

KangKyung committed:
Feat : dummy(json) 추가

- 날짜 포맷 형식에 맞게 변경
- summary 한 줄만 들어가도록 조절
- 클릭 후 제목 + 내용 삽입

KangKyung committed:
Feat: split view code base 구현

- Issue
  - 기기가 아이폰인 경우, 화면전환이 안되는 문제
  - 기기가 아이폰인 경우, 가로화면 때 split 형태가 되지않는 문제
  - 기기아 아이패드인 경우, 세로화면에서 뒤로가기 버튼이 생성되지 않는 문제

KangKyung committed:
Style: 띄어쓰기 통일 및 주석제거

KangKyung committed:
Merge pull request #1 from KangKyung/dev-step1

Dev step1

delmaSong commented on CloudNotes/CloudNotes/Model/MemoInfo.swift:
```diff
@@ -0,0 +1,25 @@
+    case lastModified = "last_modified"
+  }
+  
+  init(title: String, lastModified: Double, body: String) {
+    self.title = title
+    self.lastModified = lastModified
+    self.body = body
+  }
```
구조체에선 이런 이니셜라이저는 기본제공된답니다!

KangKyung replied:
아하..! 구조체를 자주 사용하지 않다보니 몰랐던 사실이네요!!
감사합니다😁

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,93 @@
+    return stackView
+  }()
+  
+  override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+    super.init(style: style, reuseIdentifier: reuseIdentifier)
+    
+    stackView3.addArrangedSubview(dateLabel)
+    stackView3.addArrangedSubview(summaryLabel)
+    
+    stackView2.addArrangedSubview(titleLabel)
+    stackView2.addArrangedSubview(stackView3)
+    
+    contentView.addSubview(stackView2)
+  }
+  
+  required init?(coder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
```
두 이니셜라이저의 차이는 무엇인가요?
어떨 때 fatalError가 발생하나요?
fatalError가 발생하지 않게 할 수 있을까요?

KangKyung replied:
커스텀 셀을 만들어주기 위해 이니셜라이저를 추가하는 과정에서 오류가 발생하여,
Xcode가 제안하는 대로 진행하여 만들어준 부분이었습니다.
(제대로 공부하지 않고 사용했던 것이죠..😢)

찾아본 결과, `required init`구문을 적어줘야하는 이유는
해당 클래스는 `NSCoder`프로토콜을 채택하고 있으며, 이니셜라이저를 재정의하면서
다른 이니셜라이저는 자동상속되지 않았기 때문이라는걸 알게되었습니다..!
따라서,
```swift
  required init?(coder decoder: NSCoder) {
      super.init(coder: decoder)
  }
```
이런 식으로, 호출되면 `super`의 `init`과 같은 역할을 하게끔 만들어줘야한다고 생각을 했어요.
하지만, `required init`구문이 어쩔때 호출되는 건지, 이 구문의 역할은 무엇인지는 찾지못했어요😢
혹시 해당 부분에 대해 어떤 키워드로 찾아봐야할지,
혹은 참고해보면 좋을 아티클, 공식홈페이지 등을 조언해주시면 제대로 공부해보겠습니다..!!

delmaSong replied:
[init(frame:)](https://developer.apple.com/documentation/uikit/uiview/1622488-init) 이 부분의 discussion을 읽어보시면 도움이 될듯합니다 😃

KangKyung replied:
> required init구문이 어쩔때 호출되는 건지, 이 구문의 역할은 무엇인지

코드로 뷰를 만드는 경우, 해당 이니셜라이저가 필요하다는 것을 알았습니다..!
뷰를 커스텀 하는 경우에 두 이니셜 라이저가 꼭 필요하다는 것도 알았고요!
아예 모르고 넘어갈 수 있었던 부분이었는데,
짚어주셔서 감사합니다☺️

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,93 @@
+    stackView2.frame = contentView.bounds
+  }
+  
+  func update(info: MemoInfo) {
+    titleLabel.text = info.title
+    
+    let doubleDate = info.lastModified
+    let date = Date(timeIntervalSince1970: doubleDate)
+    let dateFormatter = DateFormatter()
+    dateFormatter.locale = Locale(identifier: Locale.current.identifier)
+    dateFormatter.dateFormat = "yyyy-MM-dd"
+    dateLabel.text = dateFormatter.string(from: date)
+    
```
날짜 형식을 바꾸는 작업을 별개의 객체나 메소드에서 분리해서 작업하도록 하면 어떨까요?

KangKyung replied:
코드의 재사용성을 고려하여, 분리작업을 조언해주신게 맞을까요?
그렇다면 바로 떠오르는 방법은 메소드로 분리하는 건데,
이번에는 객체로 분리해보도록, 도전 해보겠습니다🔥🔥

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,93 @@
+    
+    let summary = info.body.components(separatedBy: ".").first
+    summaryLabel.text = summary
+    
+    self.accessoryType = .disclosureIndicator
+  }
```
이런 작업은 update 될 때보다 이후에 계속 변경되는 값이 아니니까 초기화시에 한번만 설정되도록 하는 게 좋을 것 같아요!

KangKyung replied:
앗..! 깊게생각하지 못했던 부분이에요😭
꼼꼼하게 잘 찾아주셔서 감사합니다!

delmaSong commented on CloudNotes/CloudNotes/ViewModel/ListViewModel.swift:
```diff
@@ -0,0 +1,31 @@
+
+import Foundation
+
+class ListViewModel {
+  private var memoInfoList: [MemoInfo] = []
+  
+  init() {
```
클래스 이름은 View + Model인데 여긴 모델과 관련된 작업만 있는 것 같은데요?

delmaSong commented on CloudNotes/CloudNotes/ViewModel/ListViewModel.swift:
```diff
@@ -0,0 +1,31 @@
+  private var memoInfoList: [MemoInfo] = []
+  
+  init() {
+    do {
+      // 임시 Data
+      let filePath = Bundle.main.path(forResource: "sample", ofType: "json")!
+      let data = try String(contentsOfFile: filePath).data(using: .utf8)!
+      memoInfoList = try JSONDecoder().decode([MemoInfo].self, from: data)
+    } catch {
+      print(error)
+    }
```
제약사항 중 `코드에 느낌표(!)를 사용하지 않습니다` 라는 항목을 본 것 같군요 🤔

KangKyung replied:
JSON 데이터 활용을 위해 너무 많은 시간을 쏟지 말라는 내용을 확인하고 힘을 뺐는데...
이건 제가 너무 힘뺀거 같아요😅
제대로 옵셔널 바인딩 처리 해주도록 하겠습니다..!

delmaSong commented on CloudNotes/CloudNotes/ViewModel/MemoViewModel.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+class MemoViewModel {
+  var memoInfo: MemoInfo?
+  
+  func update(model: MemoInfo?) {
+    memoInfo = model
+  }
+}
```
컨트롤러가 그냥 `MemoInfo`를 들고있는 것과 `MemoViewModel`을 들고 있도록 했을 때의 차이가 무엇인가요? MVVM 패턴을 어떤 목적에서 사용하고자 하신건지 궁금합니다

KangKyung replied:
>  MVVM 패턴을 어떤 목적에서 사용하고자 하신건지 궁금합니다

사실, 해당 프로젝트에 MVVM 패턴을 적용한 것은 패턴에대한 학습이 가장 큰 목적이었습니다.
그래서 억지로, 어떻게든 패턴적용에 초점을 두었더니, 오히려 불필요한 부분이 생긴거 같아요😢
해당 프로젝트는 이 이상의 코드확장(?)이 없을 것으로 판단되어
컨트롤러가 그냥 `MemoInfo`를 들고있도록 코드를 변경하려고 합니다..!

KangKyung replied:
``` swift
// CloudNotes/CloudNotes/View/SceneDelegate.swift
extension SceneDelegate: ListViewControllerDelegate {
  func didTapMenuItem(at index: Int, model memoInfo: MemoInfo) {
    memoVC.viewModel.update(model: memoInfo)
    memoVC.updateUI()
    ...
}
```

해당 부분에서 쓰기 위해 패턴을 활용하였습니다!
셀 클릭 이벤트가 발생하면
뷰모델을 업데이트 시킨 뒤, 해당 뷰가 업데이트되도록 구현하였습니다.
혹시 해당 부분을 더 매끄럽게 만들 수 있는 방법이 있을까요..?😭

delmaSong commented on CloudNotes/CloudNotes/View/AppDelegate.swift:
```diff
@@ -0,0 +1,48 @@
+import UIKit
+import CoreData
+
+@main
+class AppDelegate: UIResponder, UIApplicationDelegate {
+  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+    return true
+  }
+  
+  // MARK: UISceneSession Lifecycle
+  func application(_ application: UIApplication, configurationForConnecting connectingSceneSession: UISceneSession, options: UIScene.ConnectionOptions) -> UISceneConfiguration {
+    return UISceneConfiguration(name: "Default Configuration", sessionRole: connectingSceneSession.role)
+  }
+  
+  func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {}
+  
```
scene을 지원하지 않는 버전이라면 어떻게 화면을 보여줄 수 있을까요?

delmaSong commented on CloudNotes/CloudNotes/View/ListViewController.swift:
```diff
@@ -0,0 +1,71 @@
+}
+
+extension ListViewController: UITableViewDelegate {
+  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+    let memoInfo = viewModel.memoInfo(at: indexPath.row)
+    delegate?.didTapMenuItem(at: indexPath.row, model: memoInfo)
+    
+    tableView.deselectRow(at: indexPath, animated: true)
+  }
```
셀이 선택되었을 때 화면 전환의 동작이 여기서 일어나지 않는 이유가 있을까요? 
여기서 화면 전환 동작이 일어나도록 작업해주시면 고민하셨던 부분 해결에 도움이 될 것 같습니다

KangKyung replied:
현재 제가 구현한 화면은
1. 아이폰의 경우
셀 클릭 이벤트가 일어나면, 현재 화면을 `ListViewController` -> `MemoViewController` 전환시킵니다.
2. 아이패드의 경우
셀 클릭 이벤트가 일어나면, `MemoViewController`의 화면을 업데이트 시킵니다.

제가 문제를 해결하지 못했던 이유는, 말씀해주신 부분에 화면전환동작을 넣어주면
아이패드의 경우에 제대로 동작하지 않았기 때문입니다.
(사진처럼, `ListViewController` -> `MemoViewController` 전환이 됩니다)
<img width="367" alt="스크린샷 2021-06-07 오후 9 49 31" src="https://user-images.githubusercontent.com/38858863/121019386-581c9c00-c7da-11eb-8c4e-05137f94b5a8.png">
`MemoViewController`의 화면을 업데이트시켜야 했기 때문에, 말씀해주신 부분에 화면전환동작을 넣을수가 없었습니다.

생각보다 꽤 오래 생각해봤는데, 명쾌한 해결책을 찾지 못했습니다😢
해당 부분에대해 조금만 더 구체적인 조언을 구해도 될까요..?

delmaSong replied:
`UIViewController`의 프로퍼티로 존재하는 [splitViewController](https://developer.apple.com/documentation/uikit/uiviewcontroller/1623187-splitviewcontroller)를 알아보시면 해결이 될 것 같아요

KangKyung replied:
해당 부분을 `Delegate`를 이용하여
`UISplitViewController`의 `showDetailViewController`함수로 해결하였습니다!

delmaSong replied:
좋습니다 👏🏼

delmaSong commented on CloudNotes/CloudNotes/View/SceneDelegate.swift:
```diff
@@ -0,0 +1,47 @@
+//
+
+import UIKit
+
+class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+  var window: UIWindow?
+  private let splitVC = SplitViewController()
+  private let listVC = ListViewController()
+  private let memoVC = MemoViewController()
+  
+  func scene(_ scene: UIScene, willConnectTo session: UISceneSession,
+             options connectionOptions: UIScene.ConnectionOptions) {
+    guard let windowScene = (scene as? UIWindowScene) else { return }
+    window = UIWindow(windowScene: windowScene)
+    
+    listVC.title = "메모"
+    listVC.delegate = self
+    
+    splitVC.viewControllers = [
+      UINavigationController(rootViewController: listVC),
+      UINavigationController(rootViewController: memoVC)
+    ]
+    
+    window?.rootViewController = splitVC
+    window?.makeKeyAndVisible()
```
`SceneDelegate`에서는 화면에 무엇을 보여줄지 가장 근간이 되는 코드만 남도록 `rootViewController`를 지정해주는 작업만 하고 나머지 작업은 `SplitViewController` 내부에서 처리하는 게 더 자연스러울 것 같습니다

KangKyung replied:
구현에만 초점을 많이둬서 발견하지 못한 부분이었던거 같아요😢
말씀대로, `SceneDelegate`에는 화면에 무엇을 보여줄지 가장 근간이 되는 코드만 남기고
나머지 작업은 `SplitViewController`로 옮겼습니다..!

delmaSong commented on CloudNotes/CloudNotes/View/SceneDelegate.swift:
```diff
@@ -0,0 +1,47 @@
+    (UIApplication.shared.delegate as? AppDelegate)?.saveContext()
+  }
+}
+
+extension SceneDelegate: ListViewControllerDelegate {
+  func didTapMenuItem(at index: Int, model memoInfo: MemoInfo) {
+    memoVC.viewModel.update(model: memoInfo)
+    memoVC.updateUI()
+    
+    // FIXME: - 기기가 아이폰일 경우, 화면전환이 안되는 문제
+    if UIDevice.current.userInterfaceIdiom == .phone {
+      listVC.navigationController?.pushViewController(memoVC, animated: true)
+    }
+  }
+}
```
앱 전반을 관리하는 SceneDelegate가 특정 뷰에 종속되는 작업을 하는 프로토콜을 채택하는 게 부자연스러운 것 같아요. 다른 곳으로 옮겨볼 수 있을까요?

delmaSong replied:
그리고 내부에서 전달받은 index를 사용하지 않고있는데, index와 memoInfo가 함께 전달되어야 하는 이유가 무엇인가요? 둘 중 하나만 전달되어도 될 것 같아서요.

KangKyung replied:
`SceneDelegate`에는 화면에 무엇을 보여줄지 가장 근간이 되는 코드만 남기고
나머지 작업은 `SplitViewController`로 옮기는 과정에서
프로토콜 부분역시 옮겼습니다..!

KangKyung replied:
기존에 index를 사용하다가, memoInfo를 이용한 방식으로 바꿨는데
그 과정에서 삭제하지 못한 부분인거 같습니다...
이런 부분들이 자꾸 보이니까, 너무 성급하게 PR요청을 드린거 같아 굉장히 부끄럽고 죄송스러워요😢

delmaSong commented on CloudNotes/CloudNotes/View/SplitViewController.swift:
```diff
@@ -0,0 +1,21 @@
+import UIKit
+
+class SplitViewController: UISplitViewController, UISplitViewControllerDelegate {
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    self.delegate = self
+  }
+  
+  func splitViewController(_ splitViewController: UISplitViewController,
+                           collapseSecondary secondaryViewController: UIViewController,
+                           onto primaryViewController: UIViewController) -> Bool {
+    return true
```
`preferredDisplayMode` 를 비롯해 `UISplitViewController`의 속성에 대해 좀 더 알아보시면 
`해결하지 못했다고 적어주신 부분` 해결에 도움이 될 것 같습니다!

KangKyung replied:
`preferredDisplayMode = .oneBesideSecondary`를 추가하여 해당 문제를 해결하였습니다👏🏻

delmaSong commented on CloudNotes/CloudNotes/View/SplitViewController.swift:
```diff
@@ -0,0 +1,21 @@
+//
+
+import UIKit
+
+class SplitViewController: UISplitViewController, UISplitViewControllerDelegate {
```
final 키워드에 대해 알아보셔요

KangKyung replied:
주로, 재정의 할 필요가 없을 때나 클래스 자체가 상속되는 것을 막으려고 할 때 `final`키워드를 사용하는 것으로 알고있습니다.
해당되는 class에 `final`키워드를 무조건 붙여주는 것이 좋은 습관(?)일까요?
(마치, 보다 안전한 코드를 짜기위해 클래스 내부에서만 사용되는 변수, 함수를 `private`으로 해주는 것 처럼요!)

KangKyung replied:
제가 생각한게 맞다면, 해당되는 class에 모두 `final`처리를 해주는게 좋은건지도 궁금합니다..!

delmaSong replied:
네 맞습니다! final 키워드가 붙어있는 경우면 컴파일러가  상속관계를 따져보지 않아도 되고 static dispatch를 해서 성능상의 이점이 있다고 해요. [출처](https://github.com/apple/swift/blob/main/docs/OptimizationTips.rst#advice-use-final-when-you-know-the-declaration-does-not-need-to-be-overridden) 따라서 상속이 필요한 클래스가 아니라면 final 키워드는 습관처럼 붙여주는 것이 좋을 것 같습니다.

delmaSong reviewed (changes requested):
강경! 안녕하세요 😃 수고많으셨습니다!

`해결하지 못한 점` 으로 
1. 기기가 아이폰인 경우, 화면전환이 안되는 문제 
2. 기기가 아이폰인 경우, 가로화면 때 split 형태가 되지않는 문제 
3. 기기가 아이패드인 경우, 세로화면에서 뒤로가기 버튼이 생성되지 않는 문제
를 말씀해주셨는데, 해당 부분들에 대해서는 아래 코멘트를 확인해보시면 도움이 될 것 같습니다! 💪🏽

코드로 UI를 잘 구현해주셨네요! 오토레이아웃도 잘 적용이 된 것 같아요 👏🏼
코드 보면서 궁금한 점들에 대해 코멘트 달았으니 확인해주세요 🔥😃
지금 당장은 코멘트를 먼저 확인해주시고 해당 부분에 대한 수정사항이 어느정도 적용이 되었다면~
앱 첫 로드시 빈 화면이 안보이도록! (맨 상단에 있는 셀의 정보가 보이도록!?) 까지 구현해주시면 사용성에 있어서 좀 더 자연스러울 것 같아요~!
고생많으셨습니다 🏔

KangKyung committed:
Refactor: 구조체 기본이니셜라이저 삭제

KangKyung committed:
Refactor: UITableViewCell required init 수정

KangKyung committed:
Refactor: 날짜 형식 변환로직 객체로 분리

KangKyung committed:
Refactor: accessoryType정의구문 위치 이동

- accessoryType = .disclosureIndicator부분 update() -> init()이동

KangKyung committed:
Refactor: Decoding부분 optional binding처리

KangKyung committed:
Refator: 코드이동(SceneDelegate -> SplitViewController)

- SceneDelegate에서는 화면에 무엇을 보여줄지 가장 근간이 되는 코드만 남도록
  rootViewController를 지정해주는 작업만 하고
  나머지 작업은 SplitViewController 내부에서 처리하도록 변경

KangKyung committed:
Fix: 아이폰인 경우 가로화면이 split형태가 되지않는 문제

- preferredDisplayMode = .oneBesideSecondary로 해결

KangKyung committed:
Style: 변수명 변경 및 함수의 필요없는 매개변수 삭제

- 이름축약표현 변경: VC -> ViewController
- func didTapMenuItem(at index: Int, model memoInfo: MemoInfo)
  `at index: Int`삭제 -> didTapMenuItem(model memoInfo: MemoInfo)

KangKyung committed:
Merge pull request #2 from KangKyung/dev-step1

FeedBack 적용

KangKyung commented:
@delmaSong
delma가 주신 피드백대로 수정한 부분도 있고,
추가적으로 더 궁금한 부분도 있고,
아직 어떻게 변경해줘야 할지 고민중인 부분도 있습니다!

피드백을 받고, 부끄러운 부분도 많았고...
아직 부족한 부분은 더더욱 많다는 점을 느꼈어요😢
피드백들에 대한 각각의 코멘트를 확인해주시고, 한번 더 멋진 조언 부탁드리겠습니다..!

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,85 @@
+    stackView2.frame = contentView.bounds
+  }
+  
+  func update(info: MemoInfo) {
+    titleLabel.text = info.title
+    dateLabel.text = DateConvertor(date: info.lastModified).result()
+    summaryLabel.text = info.body.components(separatedBy: ".").first
+  }
```
`update` 메서드 내에서 `DateConvertor` 인스턴스를 매번 생성하기보다는 생성해둔 채로 메서드만 호출하는 게 어떨까요?

KangKyung replied:
당연한 부분이었는데 체크하지 못했어요😢
한 번만 만들고, 필요할 때 마다 함수를 호출시키는 형태로 변경하였습니다..!

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,85 @@
+    return stackView
+  }()
+  
+  override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+    super.init(style: style, reuseIdentifier: reuseIdentifier)
+    
+    stackView3.addArrangedSubview(dateLabel)
+    stackView3.addArrangedSubview(summaryLabel)
+    
+    stackView2.addArrangedSubview(titleLabel)
+    stackView2.addArrangedSubview(stackView3)
+    
+    contentView.addSubview(stackView2)
+    
+    self.accessoryType = .disclosureIndicator
+  }
+  
+  required init?(coder decoder: NSCoder) {
+      super.init(coder: decoder)
+  }
```
다른 코멘트에도 남겼지만 `init(frame:)`과 `init(coder:)`의 차이에 대해 알아보시면 좋겠습니다 😃

delmaSong commented on CloudNotes/CloudNotes/View/SplitViewController.swift:
```diff
@@ -0,0 +1,50 @@
+  }
+}
+
+extension SplitViewController: ListViewControllerDelegate {
+  func didTapMenuItem(model memoInfo: MemoInfo) {
+    memoViewController.viewModel.update(model: memoInfo)
+    memoViewController.updateUI()
+    
+    // FIXME: - 기기가 아이폰일 경우, 화면전환이 안되는 문제
+    if UIDevice.current.userInterfaceIdiom == .phone
+        && UITraitCollection.current.horizontalSizeClass == .compact {
+      listViewController.navigationController?.pushViewController(memoViewController, animated: true)
+    }
+  }
+}
```
`viewModel.update(model:)`을 호출 한 뒤 매번 반드시 `updateUI()`를 호출해줘야 하는 거라면 프로퍼티 옵저버를 이용해서 viewModel이 들고있는 모델이 변경될 때마다 등록된 어떤 동작이 매번 자동적으로 호출되게 만들면 어떨까요?

KangKyung replied:
> 프로퍼티 옵저버를 이용해서 

말씀대로, 프로퍼티 옵저버를 이용해서 어떤 것이 호출되게 만들어야하나 고민했습니다.
`memoViewModel`의 프로퍼티(`memoInfo`)가 바뀌면, UI를 업데이트시키는 동작을 해줘야 하는데
그렇게 하기 위해 `memoViewModel`은 자신을 호출한 `memoViewController`안에 있는 `updateUI`함수를 호출해야만 합니다.
억지로 어떻게든(?) 해보려면 할 수 있을거 같은데, 그 전에 혹시 다른 방법은 없는지 찾아보았습니다.

저는 처음에  `viewModel.update(model:)`안에
`escaping closure`를 이용한 `completion`함수의 호출을 떠올렸어요.
```swift
func update(of model: MemoInfo?, completion: @escaping () -> Void = {}) {
  memoInfo = model
  completion()
}
```
하지만, 뭔가 (한가지 일만 하는게 아닌)많은 일을하는 듯한 느낌(?)이 들어서 두 번째로 생각이 든게 `KVO`입니다.
그런데 `KVO`는 Objective-C 런타임에 의존하기 때문에, 순수한 swift 코드에서는 그리 좋지 않다는 글을보고 잠시 생각을 멈췄습니다.

그래서.. 정말 곰곰곰히 생각하다가 질문 드립니다!
1. 프로퍼티 옵저버를 어떻게 이용해야할지 감이 오질 않습니다😭
피드백을 받고난 뒤, 제일 많은 시간을 할애하여 고민해봤는데
어떤식으로 응용해야하는지 도저히 모르겠어요..
2. 이 문제를 해결함에 있어, 프로퍼티 옵저버가 아닌 다른 방법(`escaping closure`, `KVO`등)에 대해
delma는 어떻게 생각하는지가 궁금합니다..!

delmaSong replied:
예시로 들어주신 클로저나 KVO도 충분히 좋은 선택이 될 수 있을 것 같아요! 제가 가장 간단하게 생각했던 건 MemoViewController 내에서 
```swift 
final class MemoViewController: UIViewController { 
 let viewModel = MemoViewModel() {
  didSet {
    updateUI()
  }
}
```
처럼 하는 거였는데 이거 말고도 ViewModel의 프로퍼티로 어떤 핸들러를 가지고 있게 해서 해당 뷰모델이 변경되었을 때 그 핸들러가 실행되도록(프로퍼티 옵저버+핸들러의 조합이겠네욤)도 할 수있을 것 같구요! just do it! 고민 많이 해보신 것 같은데 일단 코드로 ㄱ ㄱ

delmaSong commented on CloudNotes/CloudNotes/View/ListViewController.swift:
```diff
@@ -0,0 +1,71 @@
+}
+
+extension ListViewController: UITableViewDelegate {
+  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+    let memoInfo = viewModel.memoInfo(at: indexPath.row)
+    delegate?.didTapMenuItem(model: memoInfo)
+    
+    tableView.deselectRow(at: indexPath, animated: true)
+  }
```
[코멘트](https://github.com/yagom-academy/ios-cloud-notes/pull/27#discussion_r648245622) 를 확인해보시면 도움이 될 것 같습니다!

delmaSong commented on CloudNotes/CloudNotes/View/ListCell.swift:
```diff
@@ -0,0 +1,85 @@
+  override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+    super.init(style: style, reuseIdentifier: reuseIdentifier)
+    
+    stackView3.addArrangedSubview(dateLabel)
+    stackView3.addArrangedSubview(summaryLabel)
+    
+    stackView2.addArrangedSubview(titleLabel)
+    stackView2.addArrangedSubview(stackView3)
```
stackView2 등 숫자로도 구분할 수 있겠지만 뭘 나타내는지 파악하는 데 비교적 어려우므로 되도록 의미를 담을 수 있는 네이밍이 더 좋을 것 같습니다 ㅎㅎ

KangKyung replied:
앗..! 다 하고 고치려고 했던 부분인데 잊어버렸어요 😭😭
서툴었던 부분을 찾아주셔서 감사합니다🙏🏼

delmaSong reviewed (commented):
댓글 남겨주신 부분과 리팩토링 하신 부분 잘 봤습니다!
지금 화면이 분할된 상태에서 제대로 뜨지 않는 문제와
뷰모델을 좀 더 매끄럽게 사용할 수 있는 방법 등에 대해 고민중이신 것 같은데
아래 코멘트 확인해주시고! 궁금하신 게 있으시면 답글 달아주시면 가능한 빨리 확인할게요!
수고많으셨어요 👏🏼 

기능적으로 제대로 동작하지 않는 부분이 있어서 스텝2로 넘어가기엔 조금 아쉬운 것 같은데,, 강경 생각은 어떠세요? 이번 코멘트들은 스텝2에서 반영하시는 게 더 나을까요?

KangKyung committed:
Fix: 기기가 아이폰일 경우, 화면전환이 안되는 문제 해결

- 화면전환방식 변경:
  navigationController?.pushViewController -> showDetailViewController

KangKyung committed:
Refactor: class final 처리

KangKyung committed:
Refactor: DateConvertor 수정

- 필요할 때 마다 오브젝트를 가져오던 방식
  -> 오브젝트를 한 번만 불러와서 함수를 호출시키는 방식 으로 변경

KangKyung committed:
Style: ListCell의 stackView 네이밍 변경

KangKyung committed:
Merge pull request #3 from KangKyung/dev-step1

FeedBack 적용

KangKyung commented:
@delmaSong 
기능적으로 제대로 동작하지 않는 부분은 얼추 마무리 하였습니다!
검토🔎해주시고, merge 부탁 드리겠습니다..!
(나머지 부분은 step2를 진행하면서 함께 해결하겠습니다!!)

그리고 delma가 피드백 해주시는 내용을 보고 코멘트를 또 남겼는데,
1. `UISplitViewController`의 `showDetailViewController`를 이용한 화면전환 문제해결이 적절한 방법이었는지
2. 프로퍼티 옵저버와 관련한 코멘트에대한 피드백

두 부분에 대해 아직 더 자세한 피드백을 받고싶어요⚠️

추가로,
- scene을 지원하지 않는 버전일 경우의 대응방법
- `init(frame:)`과 `init(coder:)`의 차이

해당 부분은 좀 더 생각해 보면서, step2를 진행하면서 해결해보고 싶습니다!
